### PR TITLE
:sparkles: Adding a Scope to the running of the rule engine

### PR DESCRIPTION
**Pull request by shawn-hurley**
The scope will allow for users to call the engine directly will be able to add information to the condition context. Providers are still responsible for restricting based on that condition context.

Working off of @jmle awesome work to get the file paths respected in providers

### CODE REVIEW

File: provider/lib.go [hunk 1]
```diff
@@ -71,9 +76,15 @@ func GetFiles(configLocation string, filepaths []string, patterns ...string) ([]
 		}
 	} else if len(filepaths) == 1 {
 		// Currently, rendering will render a list as a space separated paths as a single string.
+		log.Info("here file path 1")
 		patterns := strings.Split(filepaths[0], " ")
 		for _, pattern := range patterns {
+			if p, err := filepath.Rel(configLocation, pattern); err == nil {
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply provider/lib.go:76]: Updated to use the full file path when asking if it matches the pattern.

File: engine/scopes.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+var _ Scope = &includedPathScope{}
+
+func (i *includedPathScope) Name() string {
+	return "IncludedPathScope"
+}
+
+// This will only update conditionCTX if filepaths is not set.
+func (i *includedPathScope) AddToContext(conditionCTX *ConditionContext) error {
+	// If any chain template has the filepaths set, only use those.
```
Comment (jmle) [inline engine/scopes.go:66]: I guess in this case we are fine with scoping with the existing filepaths, is that the reason for this?

File: provider/internal/builtin/service_client.go [hunk 3]
```diff
@@ -47,16 +47,38 @@ func (p *builtinServiceClient) Evaluate(ctx context.Context, cap string, conditi
...
+					matched, err = filepath.Match(c.Pattern, path)
+					if err != nil {
+						continue
+					}
+				}
+				if matched {
+					matchingFiles = append(matchingFiles, path)
+				}
+			}
+		} else {
```
Comment (jmle) [inline provider/internal/builtin/service_client.go:77]: The problem here is that if we have a rule with filepaths and scopes are also added, its filepaths are ignored? That would essentially change the rule. Sorry if I'm not understanding this correctly.

**Review by jmle — APPROVED** [review]

**Review by jmle — CHANGES_REQUESTED** [review]
Just some questions

File: engine/scopes.go (hunk 2 above)
Comment (shawn-hurley) [reply engine/scopes.go:66]: Yeah, say for instance that you scope to pom.xml, but the condition is looking at /resources/web. in this case, we want the engine to more or less ignore this rule; we are saying in the scope that it only cares about pom.xml and only gives me results related to that file.

Does that help?

File: provider/internal/builtin/service_client.go (hunk 3 above)
Comment (shawn-hurley) [reply provider/internal/builtin/service_client.go:77]: I tried to answer this question above as well as that question, but you are basically on point. 

say a rule is searching for web.xml (hardcode or found) but when we set the scope of the analysis, we only want pom.xml. In this case I **think** the right thing to do, is only search the filepaths that is the intersection of scope of the analysis and the filepaths from the rule. 

Basically I imagine the scope, is as if you copied the included paths to $PLACE and only ran analysis on $PLACE. 

Does that help answer the question?

Comment (jmle) [reply provider/internal/builtin/service_client.go:77]: Makes sense, thanks!

**Pull request merged**